### PR TITLE
Social media

### PR DESCRIPTION
**Pull request by arek1-618**
Social media

### CODE REVIEW

File: app/SocialMedia/Twitter.php [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+	 *
+	 * @param string $message
+	 *
+	 * @throws \App\Exceptions\AppException
+	 * @throws \yii\db\Exception
+	 */
+	private function logInfoDb(string $message)
+	{
+		if (static::$logOn) {
+			static::logInfo('[' . $this->userName . ']: ' . $message);
```
Comment (rskrzypczak) [inline app/SocialMedia/Twitter.php]: ('[' . $this->userName . ']: ' . $message) = > ("[{$this->userName}]: {$message})
and other place too

File: app/SocialMedia/Base.php [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+	/**
+	 * Log to db.
+	 *
+	 * @param string $typeOfLog
+	 * @param string $message
+	 *
+	 * @throws \App\Exceptions\AppException
+	 * @throws \yii\db\Exception
+	 */
+	public static function log(string $typeOfLog, string $message)
```
Comment (mariuszkrzaczkowski) [inline app/SocialMedia/Base.php]: typeOfLog > type

Comment (mariuszkrzaczkowski) [reply app/SocialMedia/Base.php]: full logs to App\Log::log() and general information for synchronization to l_#__social_media_logs

File: app/SocialMedia/Twitter.php [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+			static::$twitterConnection->setDecodeJsonAsArray(true);
+		}
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function retrieveDataFromApi()
+	{
+		if (!\App\RequestUtil::isNetConnection()) {
```
Comment (mariuszkrzaczkowski) [inline app/SocialMedia/Twitter.php]: this should be checked before calling all functionality

File: app/SocialMedia/Twitter.php [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+			$this->logErrorDb('No internet connection');
+			return;
+		}
+		$db = \App\Db::getInstance();
+		$maxId = (new \App\Db\Query())
+			->from('u_#__social_media_twitter')
+			->where(['twitter_login' => $this->userName])
+			->max((new \yii\db\Expression('CAST(id_twitter AS INT)')));
```
Comment (mariuszkrzaczkowski) [inline app/SocialMedia/Twitter.php]: allowed in SQL 98?

File: app/SocialMedia/Twitter.php [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+		$param['screen_name'] = $this->userName;
```
Comment (mariuszkrzaczkowski) [inline app/SocialMedia/Twitter.php]: the variable is not declared

File: app/SocialMedia/Twitter.php [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+		$cnt = 0;
+		foreach ($allMessages as $rowTwitter) {
+			$rowTwitter['id'] = \App\Purifier::encodeHtml($rowTwitter['id']);
+			$rowTwitter['created_at'] = \App\Purifier::encodeHtml($rowTwitter['created_at']);
+			$rowTwitter[$rowTwitter[$indexOfText]] = \App\Purifier::encodeHtml($rowTwitter[$indexOfText]);
+			if (!isset($rowTwitter['user']['name'])) {
+				throw new \App\Exceptions\AppException('ERR_NO_VALUE||"user name"');
+			}
+			$rowTwitter['user']['name'] = \App\Purifier::encodeHtml($rowTwitter['user']['name']);
+			if (!(new \App\Db\Query())->from('u_#__social_media_twitter')->where(['id_twitter' => $rowTwitter['id']])->exists()) {
```
Comment (mariuszkrzaczkowski) [inline app/SocialMedia/Twitter.php]: this should be a function with a cache

File: app/SocialMedia/Twitter.php [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+				$db->createCommand()->insert('u_#__social_media_twitter', [
+					'id_twitter' => $rowTwitter['id'],
+					'twitter_login' => $this->userName,
+					'twitter_name' => $rowTwitter['user']['name'],
+					'message' => $rowTwitter[$indexOfText],
+					'created' => \DateTimeField::convertToUserTimeZone($rowTwitter['created_at'])->format('Y-m-d H:i:s'),
```
Comment (mariuszkrzaczkowski) [inline app/SocialMedia/Twitter.php]: App\Fields\DateTime::formatToDb() ??

File: app/SocialMedia/Twitter.php [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+	/**
+	 * Log error.
+	 *
+	 * @param string $message
+	 *
+	 * @throws \App\Exceptions\AppException
+	 * @throws \yii\db\Exception
+	 */
+	private function logErrorDb(string $message)
```
Comment (mariuszkrzaczkowski) [inline app/SocialMedia/Twitter.php]: ?

File: app/SocialMedia/Twitter.php [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+
+	/**
+	 * Log info.
+	 *
+	 * @param string $message
+	 *
+	 * @throws \App\Exceptions\AppException
+	 * @throws \yii\db\Exception
+	 */
+	private function logInfoDb(string $message)
```
Comment (mariuszkrzaczkowski) [inline app/SocialMedia/Twitter.php]: ?

File: app/SocialMedia/Twitter.php [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+					'message' => $rowTwitter[$indexOfText],
+					'created' => \DateTimeField::convertToUserTimeZone($rowTwitter['created_at'])->format('Y-m-d H:i:s'),
+				])->execute();
+				$cnt++;
+			}
+		}
+		if ($cnt > 0) {
+			$this->logInfoDb($cnt . ' new messages downloaded from');
+		} else {
+			$this->logInfoDb('No new messages');
```
Comment (mariuszkrzaczkowski) [inline app/SocialMedia/Twitter.php]: ?

File: app/SocialMedia/Twitter.php [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+					'twitter_login' => $this->userName,
+					'twitter_name' => $rowTwitter['user']['name'],
+					'message' => $rowTwitter[$indexOfText],
+					'created' => \DateTimeField::convertToUserTimeZone($rowTwitter['created_at'])->format('Y-m-d H:i:s'),
+				])->execute();
+				$cnt++;
+			}
+		}
+		if ($cnt > 0) {
+			$this->logInfoDb($cnt . ' new messages downloaded from');
```
Comment (mariuszkrzaczkowski) [inline app/SocialMedia/Twitter.php]: ?

File: app/SocialMedia/Twitter.php [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+
+	/**
+	 * Log warning.
+	 *
+	 * @param string $message
+	 *
+	 * @throws \App\Exceptions\AppException
+	 * @throws \yii\db\Exception
+	 */
+	private function logWarningDb(string $message)
```
Comment (mariuszkrzaczkowski) [inline app/SocialMedia/Twitter.php]: ?

File: app/SocialMedia/Twitter.php [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+				} else {
+					$this->logErrorDb('Twitter API unknown error');
+				}
+				throw new \App\Exceptions\AppException('ERR_API');
+			}
+			return $response;
+		} catch (\Throwable $e) {
+			$this->logErrorDb('Twitter API error: ' . $e->getMessage());
+		}
+		throw new \App\Exceptions\AppException('ERR_API');
```
Comment (mariuszkrzaczkowski) [inline app/SocialMedia/Twitter.php]: ERR_API ???

File: app/SocialMedia/Twitter.php [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+				if (\is_array($response['errors'])) {
+					foreach ($response['errors'] as $error) {
+						$errorMessage = \App\Purifier::encodeHtml($error['message']);
+						$errorCode = (int) $error['code'];
+						$this->logErrorDb('Twitter API error[code: ' . $errorCode . ']: ' . $errorMessage);
+					}
+				} else {
+					$this->logErrorDb('Twitter API unknown error');
+				}
+				throw new \App\Exceptions\AppException('ERR_API');
```
Comment (mariuszkrzaczkowski) [inline app/SocialMedia/Twitter.php]: ERR_API ???

File: cron/Social.php [hunk 15]
```diff
@@ -2,12 +2,13 @@
...
  *
+ * @package   Cron
+ *
  * @copyright YetiForce Sp. z o.o
  * @license   YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
  * @author    Arkadiusz Adach <a.adach@yetiforce.com>
  */
-$configTwitter = (new \Settings_SocialMedia_Config_Model('twitter'));
-$days = $configTwitter->get('archiving_records_number_of_days');
+$days = \App\SocialMedia::getInstance('twitter')->get('archiving_records_number_of_days');
```
Comment (mariuszkrzaczkowski) [inline cron/Social.php]: there should be some general function

File: cron/Social.php [hunk 16]
```diff
@@ -2,12 +2,13 @@
 /**
```
Comment (mariuszkrzaczkowski) [inline cron/Social.php]: wrong file name

File: layouts/basic/modules/Settings/SocialMedia/Index.tpl [hunk 17]
```diff
@@ -14,17 +14,34 @@
 					</div>
 				</div>
 			{/if}
+			{if !\vtlib\Cron::getInstance('LBL_GET_SOCIAL_MEDIA_MESSAGES')->isEnabled() }
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/SocialMedia/Index.tpl]: check both crons

File: layouts/basic/modules/Settings/SocialMedia/Index.tpl [hunk 18]
```diff
@@ -14,17 +14,34 @@
...
+						<a class="nav-link" data-toggle="tab" href="#twitter">
+							<span class="fab fa-twitter-square mr-1"></span>
 							<strong>{\App\Language::translate('LBL_TWITTER', $QUALIFIED_MODULE)}</strong>
 						</a>
 					</li>
 				</ul>
 				<div class="tab-content layoutContent py-3">
-					<div class="tab-pane active" id="twitter">
+					<div class="tab-pane active" id="logs">
+						{include file=\App\Layout::getTemplatePath('Logs.tpl', 'Settings:SocialMedia')}
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/SocialMedia/Index.tpl]: $QUALIFIED_MODULE

File: layouts/basic/modules/Settings/SocialMedia/Index.tpl [hunk 19]
```diff
@@ -14,17 +14,34 @@
...
+					</div>
+					<div class="tab-pane" id="twitter">
 						{include file=\App\Layout::getTemplatePath('Twitter.tpl', 'Settings:SocialMedia')}
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/SocialMedia/Index.tpl]: $QUALIFIED_MODULE

File: layouts/basic/modules/Settings/SocialMedia/Logs.tpl [hunk 20]
```diff
@@ -0,0 +1,32 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/SocialMedia/Logs.tpl]: removing the log panel and make a link to the general tab of logs (it is more extensive)

File: modules/Vtiger/handlers/SocialMedia.php [hunk 21]
```diff
@@ -16,6 +16,19 @@ class Vtiger_SocialMedia_Handler
 	 */
 	public function entityBeforeSave(\App\EventHandler $eventHandler)
 	{
+		$recordModel = $eventHandler->getRecordModel();
+		if (\App\SocialMedia::isEnableForModule($recordModel->getModuleName()) && !$recordModel->isNew()) {
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/handlers/SocialMedia.php]: if (!$recordModel->isNew() && \App\SocialMedia::isEnableForModule($recordModel->getModuleName()))

File: modules/Vtiger/views/Detail.php [hunk 22]
```diff
@@ -926,7 +926,7 @@ public function showOpenStreetMap($request)
 	public function showSocialMedia($request)
 	{
 		$recordModel = Vtiger_Record_Model::getInstanceById($request->getInteger('record'));
-		if (!Vtiger_SocialMedia_Model::isEnableForModule($recordModel)) {
+		if (!Vtiger_SocialMedia_Model::getInstanceByRecordModel($recordModel)->isEnableForRecord()) {
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/Detail.php]: Vtiger_SocialMedia_Model  >> \App\SocialMedia

File: modules/Accounts/models/DetailView.php [hunk 23]
```diff
@@ -113,7 +113,7 @@ public function getDetailViewRelatedLinks()
 				'linkicon' => '',
 			];
 		}
-		if (Vtiger_SocialMedia_Model::isEnableForModule($recordModel)) {
+		if (Vtiger_SocialMedia_Model::getInstanceByRecordModel($recordModel)->isEnableForRecord()) {
```
Comment (mariuszkrzaczkowski) [inline modules/Accounts/models/DetailView.php]: Vtiger_SocialMedia_Model  >> \App\SocialMedia

File: modules/Vtiger/handlers/SocialMedia.php [hunk 24]
```diff
@@ -16,6 +16,19 @@ class Vtiger_SocialMedia_Handler
...
+			$accountsToRemove = [];
+			foreach (Vtiger_SocialMedia_Model::getInstanceByRecordModel($recordModel)->getAllColumnName() as $uiType => $column) {
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/handlers/SocialMedia.php]: Vtiger_SocialMedia_Model  >> \App\SocialMedia

File: modules/Vtiger/handlers/SocialMedia.php [hunk 25]
```diff
@@ -25,7 +38,17 @@ public function entityBeforeSave(\App\EventHandler $eventHandler)
 	 *
 	 * @return bool
 	 */
-	public function entityBeforeDelete(App\EventHandler $eventHandler)
+	public function entityBeforeDelete(\App\EventHandler $eventHandler)
 	{
+		$recordModel = $eventHandler->getRecordModel();
+		if (\App\SocialMedia::isEnableForModule($recordModel->getModuleName())) {
+			$accountsToRemove = [];
+			foreach (Vtiger_SocialMedia_Model::getInstanceByRecordModel($recordModel)->getAllColumnName() as $uiType => $column) {
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/handlers/SocialMedia.php]: Vtiger_SocialMedia_Model  >> \App\SocialMedia

File: modules/Vtiger/models/SocialMedia.php [hunk 26]
```diff
@@ -34,34 +41,29 @@ public function __construct($recordModel)
...
-			return false;
-		}
-		$allFieldModel = $recordModel->getModule()->getFieldsByUiType(313);
-		foreach ($allFieldModel as $twitterField) {
-			if (!empty($recordModel->get($twitterField->getColumnName()))) {
-				return true;
+		foreach (\App\SocialMedia::ALLOWED_UITYPE as $uiType => $socialMediaType) {
+			$allFieldModel = $this->recordModel->getModule()->getFieldsByUiType($uiType);
+			foreach ($allFieldModel as $twitterField) {
+				if (!empty($this->recordModel->get($twitterField->getColumnName()))) {
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/SocialMedia.php]: getColumnName()   >>> getFieldName()

File: modules/Vtiger/models/SocialMedia.php [hunk 27]
```diff
@@ -76,41 +78,49 @@ public static function isEnableForModule($recordModel)
...
-		} else {
-			throw new \App\Exceptions\AppException('Incorrect data type in ' . $socialType);
-		}
+		$uiType = \App\SocialMedia::getUitypeFromParam($socialType)[0];
 		$socialAccount = [];
-		$allFieldModel = $this->recordModel->getModule()->getFieldsByUiType($uitype);
-		foreach ($allFieldModel as $twitterField) {
-			$val = $this->recordModel->get($twitterField->getColumnName());
+		foreach ($this->recordModel->getModule()->getFieldsByUiType($uiType) as $socialField) {
+			$val = $this->recordModel->get($socialField->getColumnName());
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/SocialMedia.php]: getColumnName() >>> getFieldName()

File: modules/Vtiger/models/SocialMedia.php [hunk 28]
```diff
@@ -76,41 +78,49 @@ public static function isEnableForModule($recordModel)
...
 			if (!empty($val) && $this->recordModel->isViewable()) {
-				$socialAccount[] = $val;
+				$socialAccount[$socialField->getColumnName()] = $val;
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/SocialMedia.php]: getColumnName() >>> getFieldName()

File: modules/Vtiger/models/SocialMedia.php [hunk 29]
```diff
@@ -76,41 +78,49 @@ public static function isEnableForModule($recordModel)
...
+	 *
+	 * @return string[]
+	 */
+	public function getAllColumnName()
+	{
+		$columnNames = [];
+		foreach (\App\SocialMedia::ALLOWED_UITYPE as $uiType => $socialMediaType) {
+			if (in_array($socialMediaType, $this->moduleConfig)) {
+				foreach ($this->recordModel->getModule()->getFieldsByUiType($uiType) as $socialField) {
+					$columnNames[$uiType] = $socialField->getColumnName();
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/models/SocialMedia.php]: getColumnName() >>> getFieldName()

Comment (mariuszkrzaczkowski) [reply modules/Vtiger/models/SocialMedia.php]: $columnNames  >> $fields

**Pull request merged**